kie-maven-plugin/src/main/java/org/kie/maven/plugin/GenerateCodeUtil.java

```diff
@@ -0,0 +1,125 @@
+ */
+
+package org.kie.maven.plugin;
+
+import java.io.File;
+import java.io.IOException;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.net.URLClassLoader;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+import org.apache.maven.artifact.Artifact;
+import org.apache.maven.artifact.DependencyResolutionRequiredException;
+import org.apache.maven.artifact.resolver.filter.CumulativeScopeArtifactFilter;
+import org.apache.maven.project.MavenProject;
+import org.kie.memorycompiler.JavaCompilerSettings;
+import org.kie.memorycompiler.JavaConfiguration;
+import org.kie.memorycompiler.KieMemoryCompiler;
+
+public class GenerateCodeUtil {
+
+    public static URLClassLoader getProjectClassLoader(MavenProject project, File outputDirectory, JavaCompilerSettings javaCompilerSettings) {
+        try {
+            Set<URL> urls = new HashSet<>();
+            for (String element : project.getCompileClasspathElements()) {
+                File file = new File(element);
+                javaCompilerSettings.addClasspath(file);
+                urls.add(file.toURI().toURL());
+            }
+
+            project.setArtifactFilter(new CumulativeScopeArtifactFilter(Arrays.asList("compile", "runtime")));
+            for (Artifact artifact : project.getArtifacts()) {
+                File file = artifact.getFile();
+                if (file != null) {
+                    javaCompilerSettings.addClasspath(file);
+                    urls.add(file.toURI().toURL());
+                }
+            }
+            urls.add(outputDirectory.toURI().toURL());
+
+            return URLClassLoader.newInstance(urls.toArray(new URL[0]), GenerateCodeUtil.class.getClassLoader());
+
+        } catch (DependencyResolutionRequiredException | MalformedURLException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    public static void compileAndWriteClasses(File targetDirectory, ClassLoader projectClassLoader, JavaCompilerSettings javaCompilerSettings, Map<String, String> classNameSourceMap, boolean dumpGeneratedSources) {
+        if (dumpGeneratedSources) {
+            dumpGeneratedSources(targetDirectory, classNameSourceMap);
+        }
+
+        Map<String, byte[]> compiledClassesMap = KieMemoryCompiler.compileNoLoad(classNameSourceMap, projectClassLoader, javaCompilerSettings, JavaConfiguration.CompilerType.ECLIPSE);
+
+        for (Map.Entry<String, byte[]> entry : compiledClassesMap.entrySet()) {
+            Path packagesDestinationPath = Paths.get(targetDirectory.getPath(), "classes", entry.getKey().replace('.', '/') + ".class");
+            writeFile(packagesDestinationPath, entry.getValue());
+        }
+    }
+
+    private static void dumpGeneratedSources(File targetDirectory, Map<String, String> classNameSourceMap) {
```

Comment:
This code is not really equivalent to the old one: now all sources are generated in `generated-sources` folder without the additional specific folder (like `drools` or `alpha-network-compiler`).
We probably need to document this change because if a user has some post postprocessor that assume sources to be there this could break.
In addition, is this path a valid maven generated-sources content? Do we need `main/src` subfolder?

Reply:
This is not meant to be equivalent to the old code. We don't need anymore to dump the generated java sources since we're compiling them in memory, so by default we're not writing them anymore. I added a parameter to optionally dumping the generated sources only for debugging purposes.

Reply:
I understand the now the logic is different, my comment is related the fact that even if user enable this flag, this doesn't reproduce the same output of the old version and secondly just to be sure that if user enable this feature the output in `target` is not broken (like that if you run again `mvn compile` without `clean` it produce an error)